### PR TITLE
Add DAE signature support

### PR DESCRIPTION
**Pull request by yizhang-yiz**
This PR addresses #1091.

new integration tests:
- test/integration/good/dae_good.stan
- test/integration/bad/variadic_dae/*.stan

#### Submission Checklist

- [x] Run unit tests
- Documentation
    - [x] If a user-facing facing change was made, the documentation PR is here: https://github.com/stan-dev/docs/pull/467
    - [ ] OR, no user-facing changes were made

## Release notes
Support Differential-Algebraic Equation solver.

## Copyright and Licensing

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the BSD 3-clause license (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

**Comment by WardBrian** [comment]
Do you have any sense of a timeline on https://github.com/stan-dev/math/pull/2644? 

#1027 is nearing completion, and whichever one of these merges first will create problems for the other (the typechecking and code gen of variadic functions is more or less completely rewritten in #1027)

**Comment by yizhang-yiz** [comment]
> Do you have any sense of a timeline on [stan-dev/math#2644](https://github.com/stan-dev/math/pull/2644)?

From the initial feedback from @charlesm93 it should be soon. Of course the final decision is his.

File: src/frontend/Semantic_error.ml
Comment (WardBrian) [inline src/frontend/Semantic_error.ml:38]: I would personally rather not create a new error type here and instead just make the current `IllTypedVariadicODE` into `IllTypedVariadicFn`. The type will need to include the return type (which will be either                       Stan_math_signatures.variadic_dae_fun_return_type or                       Stan_math_signatures.variadic_ode_fun_return_type) but it saves the duplication here

**Review by WardBrian — COMMENTED** [review]
One minor comment, otherwise LGTM

File: src/frontend/Semantic_error.ml
Comment (yizhang-yiz) [reply src/frontend/Semantic_error.ml]: Thanks. I just changed replace both with `IllTypedVariadicDE` with an additional arg for return type.

File: src/middle/Stan_math_signatures.ml
Comment (WardBrian) [inline src/middle/Stan_math_signatures.ml:226]:
```suggestion
```

File: src/middle/Stan_math_signatures.ml
Comment (WardBrian) [inline src/middle/Stan_math_signatures.ml:160]:
```suggestion
```

File: src/middle/Stan_math_signatures.ml
Comment (WardBrian) [inline src/middle/Stan_math_signatures.ml:233]:
```suggestion
```

**Review by WardBrian — CHANGES_REQUESTED** [review]
I think these comments in `Stan_math_signatures` provide more clutter than illumination.


Other than that, LGTM. I will approve/merge when the stan::math PR is merged and docs are approved

**Comment by WardBrian** [comment]
Looks like you'll need to update to the new pattern introduced in #1004  - should be pretty easy to see based on the variadic ode, again

**Review by WardBrian — APPROVED** [review]
Should be good to go after the math branch is merged

**Comment by rok-cesnovar** [comment]
Tests pass so feel free to merge.

**Comment by WardBrian** [comment]
Doc is under review still, but I think it will definitely get in before the release so we can merge

**Pull request merged**
